prow/sidecar/run.go

```diff
@@ -58,91 +63,75 @@ func (o Options) Run() error {
+		returnCode, err := opt.WaitForMarker(ctx, opt.MarkerFile)
+		passed = passed && err == nil && returnCode == 0
+		aborted = aborted || returnCode == 130
 	}
 
+	cancel()
 	// If we are being asked to terminate by the kubelet but we have
 	// seen the test process exit cleanly, we need a chance to upload
 	// artifacts to GCS. The only valid way for this program to exit
-	// after a SIGINT or SIGTERM in this situation is to finish]
+	// after a SIGINT or SIGTERM in this situation is to finish
 	// uploading, so we ignore the signals.
 	signal.Ignore(os.Interrupt, syscall.SIGTERM)
 
-	passed := false
-	aborted := false
-	returnCodeData, err := ioutil.ReadFile(o.WrapperOptions.MarkerFile)
-	if err != nil {
-		logrus.WithError(err).Warn("Could not read return code from marker file")
-	} else {
-		returnCode, err := strconv.Atoi(strings.TrimSpace(string(returnCodeData)))
+	errors := map[string]error{}
+	metadata := map[string]interface{}{}
+	var readers []io.Reader
+	for i, opt := range entries {
+		ent := nameEntry(i, opt)
+		if len(entries) > 1 {
+			readers = append(readers, strings.NewReader(fmt.Sprintf("==== start of %s ===\n", ent)))
+		}
+		log, err := os.Open(opt.ProcessLog)
 		if err != nil {
-			logrus.WithError(err).Warn("Failed to parse process return code")
+			logrus.WithError(err).Errorf("Failed to open %s", opt.ProcessLog)
+			readers = append(readers, strings.NewReader(fmt.Sprintf("Failed to open %s: %v", opt.ProcessLog, err)))
+		} else {
+			readers = append(readers, log)
+		}
+		metadataFile := o.WrapperOptions.MetadataFile
+		if _, err := os.Stat(metadataFile); err != nil {
+			if !os.IsNotExist(err) {
+				logrus.WithError(err).Errorf("Failed to stat %s", metadataFile)
+				errors[ent] = err
+			}
+			continue
 		}
-		passed = returnCode == 0 && err == nil
-		aborted = returnCode == 130
-	}
 
-	metadataFile := o.WrapperOptions.MetadataFile
-	if _, err := os.Stat(metadataFile); err != nil {
-		if !os.IsNotExist(err) {
-			logrus.WithError(err).Errorf("Failed to stat %s", metadataFile)
+		metadataRaw, err := ioutil.ReadFile(metadataFile)
+		if err != nil {
+			logrus.WithError(err).Errorf("cannot read %s", metadataFile)
+			errors[ent] = err
+			continue
 		}
-		return o.doUpload(spec, passed, aborted, nil)
-	}
 
-	metadataRaw, err := ioutil.ReadFile(metadataFile)
-	if err != nil {
-		logrus.WithError(err).Errorf("cannot read %s", metadataFile)
-		return o.doUpload(spec, passed, aborted, nil)
-	}
+		piece := map[string]interface{}{}
+		if err := json.Unmarshal(metadataRaw, &piece); err != nil {
+			logrus.WithError(err).Errorf("Failed to unmarshal %s", metadataFile)
+			errors[ent] = err
+			continue
+		}
 
-	metadata := map[string]interface{}{}
-	if err := json.Unmarshal(metadataRaw, &metadata); err != nil {
-		logrus.WithError(err).Errorf("Failed to unmarshal %s", metadataFile)
-		return o.doUpload(spec, passed, aborted, nil)
+		for k, v := range piece {
+			metadata[k] = v // TODO(fejta): consider deeper merge
```

Comment:
I think k8s utils has this somewhere

Reply:
Yeah hence the todo, can deal with it later. I want some logs more than I care about metadata.json correctness right now (since zero build jobs use this feature)